larpandoradlcontent/LArDeepLearning/DeepLearningTrackShowerIdAlgorithm.cc

```diff
@@ -0,0 +1,352 @@
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pMCParticleList));
+
+        const bool isU(pCaloHitList->front()->GetHitType() == TPC_VIEW_U ? true : false);
+        const bool isV(pCaloHitList->front()->GetHitType() == TPC_VIEW_V ? true : false);
+        const bool isW(pCaloHitList->front()->GetHitType() == TPC_VIEW_W ? true : false);
+
+        if (!isU && !isV && !isW) return STATUS_CODE_NOT_ALLOWED;
+
+        std::string trainingOutputFileName(m_trainingOutputFile);
+
+        if (isU) trainingOutputFileName += "_CaloHitListU.csv";
+        else if (isV) trainingOutputFileName += "_CaloHitListV.csv";
+        else if (isW) trainingOutputFileName += "_CaloHitListW.csv";
+
+        LArMCParticleHelper::PrimaryParameters parameters;
+        // Only care about reconstructability with respect to the current view, so skip good view check
+        parameters.m_minHitsForGoodView = 0;
+        // Turn off max photo propagation for now, only care about killing off daughters of neutrons
+        parameters.m_maxPhotonPropagation = std::numeric_limits<float>::max();
+        LArMCParticleHelper::MCContributionMap targetMCParticleToHitsMap;
+        LArMCParticleHelper::SelectReconstructableMCParticles(pMCParticleList, pCaloHitList, parameters,
+            LArMCParticleHelper::IsBeamNeutrinoFinalState, targetMCParticleToHitsMap);
+
+        LArMvaHelper::MvaFeatureVector featureVector;
+        for (const CaloHit *pCaloHit : *pCaloHitList)
+        {
+            int tag{TRACK};
+            float inputEnergy{0.f};
+
+            try
+            {
+                const MCParticle *const pMCParticle(MCParticleHelper::GetMainMCParticle(pCaloHit));
+                // Throw away non-reconstructable hits
+                if (targetMCParticleToHitsMap.find(pMCParticle) == targetMCParticleToHitsMap.end())
+                    continue;
+                if(LArMCParticleHelper::IsDescendentOf(pMCParticle, 2112))
+                    continue;
+                inputEnergy = pCaloHit->GetInputEnergy();
+                if (inputEnergy < 0)
+                    continue;
+
+                const int pdg{std::abs(pMCParticle->GetParticleId())};
+                if (pdg == 11 || pdg == 22)
+                    tag = SHOWER;
+                else
+                    tag = TRACK;
+            }
+            catch (...)
+            {
+                continue;
+            }
+
+            featureVector.push_back(static_cast<double>(pCaloHit->GetPositionVector().GetX()));
+            featureVector.push_back(static_cast<double>(pCaloHit->GetPositionVector().GetZ()));
+            featureVector.push_back(static_cast<double>(tag));
+            featureVector.push_back(static_cast<double>(inputEnergy));
+        }
+        // Add number of hits to end of vector than rotate (more efficient than direct insert at front)
+        featureVector.push_back(static_cast<double>(featureVector.size() / 4));
```

Comment:
Robust enough to only static cast the temporary (temporary itself won't have problems, without casting numerator)? Presumably fine, but often good to make sure all OK explicitly.

Reply:
The temporary here should be a prvalue, so it should be treated in exactly the same way as the return of non-reference return values from functions.